src/node_native_module.cc

```diff
@@ -174,6 +174,64 @@ MaybeLocal<Function> NativeModuleLoader::CompileAsModule(
   return LookupAndCompile(context, id, &parameters, result);
 }
 
+#ifdef NODE_BUILTIN_MODULES_PATH
+static std::string OnDiskFileName(const char* id) {
+  std::string filename = NODE_BUILTIN_MODULES_PATH;
+  filename += "/";
+
+  if (memcmp(id, "internal/deps", strlen("internal/deps")) == 0) {
+    id += strlen("internal/");
+  } else {
+    filename += "lib/";
+  }
+  filename += id;
+  filename += ".js";
+
+  return filename;
+}
+#endif  // NODE_BUILTIN_MODULES_PATH
+
+MaybeLocal<String> NativeModuleLoader::LoadBuiltinModuleSource(Isolate* isolate,
+                                                               const char* id) {
+#ifdef NODE_BUILTIN_MODULES_PATH
+  std::string filename = OnDiskFileName(id);
+
+  uv_fs_t req;
+  uv_file file =
+      uv_fs_open(nullptr, &req, filename.c_str(), O_RDONLY, 0, nullptr);
+  CHECK_GE(req.result, 0);
+  uv_fs_req_cleanup(&req);
+
+  std::shared_ptr<void> defer_close(nullptr, [file](...) {
+    uv_fs_t close_req;
+    CHECK_EQ(0, uv_fs_close(nullptr, &close_req, file, nullptr));
+    uv_fs_req_cleanup(&close_req);
+  });
+
+  std::string contents;
+  char buffer[4096];
+  uv_buf_t buf = uv_buf_init(buffer, sizeof(buffer));
+
+  while (true) {
+    const int r =
+        uv_fs_read(nullptr, &req, file, &buf, 1, contents.length(), nullptr);
+    CHECK_GE(req.result, 0);
+    uv_fs_req_cleanup(&req);
+    if (r <= 0) {
+      break;
+    }
+    contents.append(buf.base, r);
+  }
```

Comment:
There's still enough common functionality to factor out the read-to-std::string bit. The fact alone that you introduced a file descriptor leak in the first iteration is IMO enough reason to do it.

I suggest sticking with InternalModuleReadJSON's read-to-std::vector and return the actual bytes read so you don't have to shrink-to-fit the std::vector.

Possible future optimization: `std::move` the vector to a `v8::String::ExternalOneByteStringResource` derivative when creating the `v8::String` when it's definitely known to contain only ASCII or Latin-1.

Reply:
They have different error handling behaviour, i'm really not sure how i'm supposed to combine them.

Reply:
I'm not sure what you mean. `r < 0` vs. `r <= 0` w.r.t. the return value of `uv_fs_read()`? The second one is the (superfluous) extra read I mentioned.

Reply:
I mean that all three have different behaviour to perform when an error happens. if you have some design in mind I'll gladly accept a patch.

Reply:
I'm still unsure what you mean. The only material difference I see is that your version CHECKs when `uv_fs_open()` fails but I assume that's not what you mean.

(I don't think that's a particular good way to handle file-not-found but since this is a core developer-only feature that I have no particular interest in using myself, I don't really care. :-))

Reply:
InternalModuleReadJSON returns `undefined` to the caller if reading fails, the one in module_wrap will best-effort return all the data it managed to read (if any), and the one I'm adding asserts that none of the operations fail. Perhaps aligning these more is best left to a separate PR?

Reply:
Ah, okay. I don't think that really changes the inner open/read/close logic - you just pass the error to the caller and let it decide what to do. But if you pinky promise to follow up with another PR, I'll let it drop. :-)

Reply:
pinky promise :D